Casenorm tests: enable backslash interpretation

gmelikov opened the pull request:
`echo` must have `-e` argument to properly
interpret unusual symbols.

Signed-off-by: George Melikov <mail@gmelikov.ru>

<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Fix #7633

### Description
<!--- Describe your changes in detail -->
Thanks to @ikozhukhov , the problem resolves with this patch.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
WIP, waiting for buildbot successfull finish.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

gmelikov commented:
Ah, the problem is with non-existing `zlook` at Linux, will try to replace it.

behlendorf commented:
Thanks for looking in to this.  As part of the final fix you're going to want to remove the `casenorm` exceptions from `tests/test-runner/bin/zts-report.py` too.  Also if you'd like to iterate more quickly with the CI on this, I'd suggest adding `TEST_ZFSTESTS_TAGS="casenorm"` to your commit message so only that test group is run.

rlaager commented:
`echo -e` is not POSIX compliant. Starting from bash:
```
$ echo -e 'F\0303\0257L\0303\0253N\0303\0204m\0303\0253'
FïLëNÄmë
$ dash
$ echo -e 'F\0303\0257L\0303\0253N\0303\0204m\0303\0253'
-e FïLëNÄmë
```

Is this guaranteed to run under `bash`?

gmelikov commented:
@ikozhukhov @rlaager I've got it working via printf, just replaced lines with needed format for printf.

@behlendorf thank you, removed tests from exclusion list. 

Used `test -f` as a cheap replacement for `zlook` (maybe it's not the best path). Got a problem with some tests, will dig on weekend.

Ornias1993 commented:
@gmelikov looks like a simple (partial) fix, any news on this? Being WIP and such...

gmelikov commented:
@Ornias1993 unfortunately it's not so simple - printf does work in Linux, but some tests are failing, I think there are some problems with ksh in Linux, but I didn't find the root cause yet.

I've updated PR and excluded problematic tests from PR for now, let's see tests and maybe merge this change as is.

behlendorf commented:
This could be an issue with `ksh-2020.0.0`, or perhaps an intentional change in behavior.  When running `ksh` interactively it correctly interprets the unicode, but for some reason when invoked via the test suite not all `\uHHHH` values are interpreted (i.e \u0308).

```
ksh --version
  version         sh (AT&T Research) 2020.0.0
[behlendo@localhost zfs]$ ksh
$ printf '\u0066\u0069\u0308\u006c\u0065\u0308\u006e\u0061\u0308\u006d\u0065\u0308'
fïlënämë$ 
```

For example, a snipet from the ZTS output:

```
printf '\u0066\u0069\u0308\u006c\u0065\u0308\u006e\u0061\u0308\u006d\u0065\u0308'
fi\u0308le\u0308na\u0308me\u0308
```

Using `bash -c "-printf '...'"` as a workaround worked for me on Fedora 31.  It's pretty ugly, but we could probably live with it as a Linux-only change.  When applied all of the casenorm tests pass.  

```
NAME_C_ORIG=$(bash -c "printf '\u0046\u00ef\u004c\u00eb\u004e\u00c4\u006d\u00eb'")
NAME_C_UPPER=$(bash -c "printf '\u0046\u00cf\u004c\u00cb\u004e\u00c4\u004d\u00cb'")
NAME_C_LOWER=$(bash -c "printf '\u0066\u00ef\u006c\u00eb\u006e\u00e4\u006d\u00eb'")
NAME_D_ORIG=$(bash -c "printf '\u0046\u0069\u0308\u004c\u0065\u0308\u004e\u0041\u0308\u006d\u0065\u0308'")
NAME_D_UPPER=$(bash -c "printf '\u0046\u0049\u0308\u004c\u0045\u0308\u004e\u0041\u0308\u004d\u0045\u0308'")
NAME_D_LOWER=$(bash -c "printf '\u0066\u0069\u0308\u006c\u0065\u0308\u006e\u0061\u0308\u006d\u0065\u0308'")
```

behlendorf commented:
Ahh, it seems the locale is being lost from the environment.  @gmelikov adding the following hunk to this PR should be enough.

```diff
diff --git a/tests/zfs-tests/tests/functional/casenorm/casenorm.cfg b/tests/zfs-tests/tests/functional/casenorm/casenorm.cfg
index e35a726e7df..f955d33cd19 100644
--- a/tests/zfs-tests/tests/functional/casenorm/casenorm.cfg
+++ b/tests/zfs-tests/tests/functional/casenorm/casenorm.cfg
@@ -17,6 +17,9 @@
 # Copyright (c) 2016 by Delphix. All rights reserved.
 #
 
+LANG=en_US.UTF-8
+LC_ALL=en_US.UTF-8
+
 NAME_C_ORIG=$(printf '\u0046\u00ef\u004c\u00eb\u004e\u00c4\u006d\u00eb')
 NAME_C_UPPER=$(printf '\u0046\u00cf\u004c\u00cb\u004e\u00c4\u004d\u00cb')
 NAME_C_LOWER=$(printf '\u0066\u00ef\u006c\u00eb\u006e\u00e4\u006d\u00eb')
```

rlaager commented:
Instead of "en_US", does "C" or "C.UTF-8" work? If so, it might be better to use that. Otherwise, en_US.UTF-8 is probably fine.

gmelikov commented:
@behlendorf thank you! Unfortunately, locale fix didn't work for buildbot, so I've used bash hack.

gmelikov commented:
Bash hack didn't work either for buildbot. Interesting.

behlendorf commented:
Can you try the following.  @rlaager's suggestion works for me locally.

```
export LANG="C.UTF-8"
export LC_ALL="C.UTF-8"
```

Note that with this change I still do see 6 local failures like with the original PR.  But everything looks to be escaped properly so those failures will need to be inspected individually.

```
$ ./scripts/zfs-tests.sh -T casenorm
...
Results Summary
PASS	  13
FAIL	   6

Running Time:	00:00:45
Percent passed:	68.4%
Log directory:	/var/tmp/test_results/20191118T093144

Tests with results other than PASS that are expected:
    FAIL casenorm/mixed_formd_delete (https://github.com/zfsonlinux/zfs/issues/7633)
    FAIL casenorm/mixed_formd_lookup (https://github.com/zfsonlinux/zfs/issues/7633)
    FAIL casenorm/mixed_formd_lookup_ci (https://github.com/zfsonlinux/zfs/issues/7633)
    FAIL casenorm/mixed_none_lookup_ci (https://github.com/zfsonlinux/zfs/issues/7633)
    FAIL casenorm/sensitive_formd_delete (https://github.com/zfsonlinux/zfs/issues/7633)
    FAIL casenorm/sensitive_formd_lookup (https://github.com/zfsonlinux/zfs/issues/7633)
```

gmelikov commented:
@behlendorf ah, I thought you had all tests passed. 

Added locale fix for fedora, excluded problematic tests. We may merge this PR and tackle with other tests next time.
